### PR TITLE
kubernetes-deployment-ng

### PR DESCRIPTION
**Pull request by svenwltr**
**Roadmap**

* [x] bootstrap the project
* [x] setup CLI
* [x] get settings from cloud-infrastructure project (directly via API)
* [x] get templates from GitHub (directly via API)
* [x] fetch branch information
* [x] render manifest templates
* [x] load rendered manifests into structs
* [x] add cache for http requests
* [x] apply patch to Kubernetes (~~directly via API~~)
* [ ] support multiple contexts (production, staging)
* [ ] ???
* [ ] profit

There is still the possibility to change the project name.

Also the name `settings` for the file in cloud-infrastructure feels a bit ambiguous. Maybe someone has a better idea.

### CODE REVIEW

**Comment by svenwltr** [comment]
Output, so far:

```
% ./kubernetes-deployment deploy -v grafana
INFO[0000] kubernetes-deployment started                 Commit=ac59d49f53370b556eef66cbec6bf47fada56410 Date="Sun May 21 23:51:36 CEST 2017" Version=v4.0.0.19.gac59d49.dirty
DEBU[0000] config loaded                                 Filename="/home/sven/.rebuy/kubernetes-deployment/services.yaml" GitHubToken=4346a4dcf268f6c9a1dd4d242e208b8cefcf3841 Kubeconfig=
DEBU[0000] loaded service file                           Filename="/home/sven/.rebuy/kubernetes-deployment/services.yaml" ServiceCount=98
INFO[0000] deploying grafana/master                     
DEBU[0000] service found                                 Location=map[Owner:rebuy-de Repo:cloud-infrastructure Branch:master Path:grafana/] Name=grafana TemplateValues=map[clusterStage:production clusterDomain:production.cloud.rebuy.loc gfDatabaseType:sqlite3 gfDatabaseName:grafana]
DEBU[0000] downloading directory from GitHub             Branch=master Owner=rebuy-de Path="grafana/" Repo=cloud-infrastructure
DEBU[0001] found files in directory                      Files=4 RateLimit=5000 RateRemaining=4952 RateReset=2017-05-22 00:17:58 +0200 CEST
DEBU[0001] downloading file from GitHub                  Branch=master Owner=rebuy-de Path="grafana/deployment.yaml" Repo=cloud-infrastructure
DEBU[0002] found file                                    RateLimit=5000 RateRemaining=4951 RateReset=2017-05-22 00:17:58 +0200 CEST Size=1532 URL="https://github.com/rebuy-de/cloud-infrastructure/blob/master/grafana/deployment.yaml"
DEBU[0002] downloading file from GitHub                  Branch=master Owner=rebuy-de Path="grafana/ingress.yaml" Repo=cloud-infrastructure
DEBU[0002] found file                                    RateLimit=5000 RateRemaining=4950 RateReset=2017-05-22 00:17:58 +0200 CEST Size=332 URL="https://github.com/rebuy-de/cloud-infrastructure/blob/master/grafana/ingress.yaml"
DEBU[0002] downloading file from GitHub                  Branch=master Owner=rebuy-de Path="grafana/service.yaml" Repo=cloud-infrastructure
DEBU[0002] found file                                    RateLimit=5000 RateRemaining=4949 RateReset=2017-05-22 00:17:58 +0200 CEST Size=260 URL="https://github.com/rebuy-de/cloud-infrastructure/blob/master/grafana/service.yaml"
DEBU[0002] downloading file from GitHub                  Branch=master Owner=rebuy-de Path="grafana/volume-claim.yml" Repo=cloud-infrastructure
DEBU[0003] found file                                    RateLimit=5000 RateRemaining=4948 RateReset=2017-05-22 00:17:58 +0200 CEST Size=260 URL="https://github.com/rebuy-de/cloud-infrastructure/blob/master/grafana/volume-claim.yml"
```

**Comment by svenwltr** [comment]
Output so far:

```
% ./kubernetes-deployment generate blue-web -v | jq             
INFO[0000] kubernetes-deployment started                 Commit=dddd1b5e92cb5e3ec2d278512a171ebd3d9e321b Date="Tue Jun  6 17:27:58 CEST 2017" Version=v4.0.0.19.gdddd1b5
DEBU[0000] config loaded                                 Filename="/home/sven/.rebuy/kubernetes-deployment/services.yaml" GitHubToken="b1e2****" Kubeconfig=
DEBU[0000] loaded service file                           Filename="/home/sven/.rebuy/kubernetes-deployment/services.yaml" ServiceCount=98
INFO[0000] deploying project                             Branch=master Project=blue-web
DEBU[0000] project found                                 Location=map[Owner:rebuy-de Repo:blue-web Path:deployment/k8s/ Ref:master] Name=blue-web TemplateValues=map[gfDatabaseName:grafana clusterStage:production clusterDomain:production.cloud.rebuy.loc gfDatabaseType:sqlite3]
DEBU[0000] fetched branch information                    Author=Spea Branch=master Date=2017-05-30 22:32:04 +0000 UTC Owner=rebuy-de RateLimit=5000 RateRemaining=4999 RateReset=2017-06-06 18:28:46 +0200 CEST Ref=master Repo=blue-web SHA=84472f0cb684da77b93659a514d626f602c26d20
INFO[0000] latest commit:

commit: 84472f0cb684da77b93659a514d626f602c26d20
Author: Spea
Date:   2017-05-30 22:32:04 +0000 UTC

Deactivate jenkins and travis
 
DEBU[0000] downloading directory from GitHub             Owner=rebuy-de Path="deployment/k8s/" Ref=84472f0cb684da77b93659a514d626f602c26d20 Repo=blue-web
DEBU[0001] found files in directory                      Files=3 RateLimit=5000 RateRemaining=4997 RateReset=2017-06-06 18:28:46 +0200 CEST
DEBU[0001] downloading file from GitHub                  Owner=rebuy-de Path="deployment/k8s/service.yaml" Ref=84472f0cb684da77b93659a514d626f602c26d20 Repo=blue-web
DEBU[0001] downloading file from GitHub                  Owner=rebuy-de Path="deployment/k8s/deployment.yaml" Ref=84472f0cb684da77b93659a514d626f602c26d20 Repo=blue-web
DEBU[0001] downloading file from GitHub                  Owner=rebuy-de Path="deployment/k8s/ingress.yaml" Ref=84472f0cb684da77b93659a514d626f602c26d20 Repo=blue-web
DEBU[0001] found file                                    RateLimit=5000 RateRemaining=4996 RateReset=2017-06-06 18:28:46 +0200 CEST Size=273 URL="https://github.com/rebuy-de/blue-web/blob/84472f0cb684da77b93659a514d626f602c26d20/deployment/k8s/service.yaml"
DEBU[0001] found file                                    RateLimit=5000 RateRemaining=4995 RateReset=2017-06-06 18:28:46 +0200 CEST Size=2012 URL="https://github.com/rebuy-de/blue-web/blob/84472f0cb684da77b93659a514d626f602c26d20/deployment/k8s/deployment.yaml"
DEBU[0001] found file                                    RateLimit=5000 RateRemaining=4994 RateReset=2017-06-06 18:28:46 +0200 CEST Size=836 URL="https://github.com/rebuy-de/blue-web/blob/84472f0cb684da77b93659a514d626f602c26d20/deployment/k8s/ingress.yaml"
DEBU[0001] collected template values                     Values=map[gitBranchName:master gitCommitID:84472f0cb684da77b93659a514d626f602c26d20 clusterStage:production clusterDomain:production.cloud.rebuy.loc gfDatabaseType:sqlite3 gfDatabaseName:grafana]
{
  "kind": "Deployment",
  "apiVersion": "extensions/v1beta1",
  "metadata": {
    "name": "blue-web-master",
    "creationTimestamp": null,
    "labels": {
      "app": "blue-web",
      "branch": "master",
      "team": "platform",
      "tier": "frontend"
    }
--- snip ---
```

**Comment by svenwltr** [comment]
FYI: This already works:

```
./kubernetes-deployment generate kubernetes-status | kubectl apply -f -
```

**Comment by svenwltr** [comment]
@rebuy-de/prp-kubernetes-deployment This now has the same features as the old version. Therefore it is ready for merge. Please review.

File: pkg/api/init.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package api
+
+import (
```
Comment (bjoernhaeuser) [inline pkg/api/init.go]: what is this file for?

File: pkg/api/generate.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+			continue
+		}
+
+		obj, _, err := decode([]byte(data), nil, nil)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to decode file '%s'", name)
+		}
+
+		objects = append(objects, obj)
+
```
Comment (bjoernhaeuser) [inline pkg/api/generate.go]: one lb too much

**Comment by svenwltr** [comment]
done

**Comment by bjoernhaeuser** [comment]
Went through it, makes sense to me. Could not find anything.

File: pkg/api/init.go (hunk 1 above)
Comment (svenwltr) [inline pkg/api/init.go]: For decoding of yaml files: https://stackoverflow.com/questions/44306554/how-to-deserialize-kubernetes-yaml-file

Comment (bjoernhaeuser) [inline pkg/api/init.go]: Lets add a comment :)

File: pkg/api/generate.go (hunk 2 above)
Comment (svenwltr) [inline pkg/api/generate.go]: removed

File: pkg/api/init.go (hunk 1 above)
Comment (svenwltr) [inline pkg/api/init.go]: added

**Pull request merged**